flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CastFunctionITCase.java

```diff
@@ -18,176 +18,905 @@
+                        // https://issues.apache.org/jira/browse/FLINK-21456 Not supported currently
+                        // ARRAY
+                        // MULTISET
+                        // MAP
+                        // ROW
+                        // RAW
+                        .build(),
+                CastTestSpecBuilder.test(BOOLEAN(), "To BOOLEAN")
+                        .testCase(CHAR(3), "foo", null)
+                        .testCase(CHAR(4), "true", true)
+                        .testCase(VARCHAR(5), "FalsE", false)
+                        .testCase(STRING(), "Apache Flink", null)
+                        .testCase(STRING(), "TRUE", true)
+                        .testCase(STRING(), "", null)
+                        .testCase(BOOLEAN(), true, true)
+                        .testCase(BOOLEAN(), false, false)
+                        // Not supported - no fix
+                        // BINARY
+                        // VARBINARY
+                        // BYTES
+                        .testCase(DECIMAL(4, 3), 9.87, true)
+                        .testCase(DECIMAL(2, 1), 0.0, false)
+                        .testCase(TINYINT(), -125, true)
+                        .testCase(TINYINT(), 0, false)
+                        .testCase(SMALLINT(), 32767, true)
+                        .testCase(SMALLINT(), 0, false)
+                        .testCase(INT(), -12345678, true)
+                        .testCase(INT(), 0, false)
+                        .testCase(BIGINT(), 1234567891234L, true)
+                        .testCase(BIGINT(), 0, false)
+                        .testCase(FLOAT(), -123.456, true)
+                        .testCase(FLOAT(), 0, false)
+                        .testCase(DOUBLE(), 12345.67890, true)
+                        .testCase(DOUBLE(), 0, false)
+                        // Not supported - no fix
+                        // DATE
+                        // TIME
+                        // TIMESTAMP
+                        // TIMESTAMP_WITH_TIME_ZONE
+                        // TIMESTAMP_LTZ
+                        // INTERVAL(YEAR())
+                        // INTERVAL(DAY())
+                        // ARRAY
+                        // MULTISET
+                        // MAP
+                        // ROW
+                        // RAW
+                        .build(),
+                CastTestSpecBuilder.test(BINARY(2), "To BINARY")
+                        .testCase(CHAR(3), "foo", new byte[] {102, 111, 111})
+                        .testCase(VARCHAR(5), "Flink", new byte[] {70, 108, 105, 110, 107})
+                        // https://issues.apache.org/jira/browse/FLINK-24419 - not trimmed to 2
+                        // bytes
+                        .testCase(STRING(), "Apache", new byte[] {65, 112, 97, 99, 104, 101})
+                        // Not supported - no fix
+                        // BOOLEAN
```

Comment:
for the unsupported cases, did you check that the exception came from the validation layer instead of the code generator?

Reply:
Yep, There is this case of casting from string types with invalid values to `TIMESTAMP_LTZ` which cause NPEs which is marked with the corresponding jira issue.